quoted-strings: add allow-quoted-quotes option

rgeraskin opened the pull request:
This PR implements ~~`skip-quoted-quote`~~ `allow-quoted-quotes` option that allows strings like:
- `'foo"bar'` on `quote-type: double`
- `"foo'bar"` on `quote-type: single`

Closes use-case with quotes inside the string from #424. But still no solution for escape sequences.

adrienverge commented on yamllint/rules/quoted_strings.py:
```diff
@@ -139,6 +166,8 @@ def VALIDATE(conf):
         return 'cannot use both "required: true" and "extra-required"'
     if conf['required'] is False and len(conf['extra-allowed']) > 0:
         return 'cannot use both "required: false" and "extra-allowed"'
+    if conf['skip-quoted-quote'] is True and conf['quote-type'] == 'any':
+        return '"skip-quoted-quote" has no effect with "quote-type: any"'
```
I fear it can be unfriendly to the user to fail on that. Maybe we can just silently ignore `skip-quoted-quote` in that case, and run yamllint normally?

rgeraskin replied:
Sounds reasonable. I've removed this check.

adrienverge commented on yamllint/rules/quoted_strings.py:
```diff
@@ -112,6 +115,28 @@
     - "localhost"
     - this is a string that needs to be QUOTED
+
+#. With ``quoted-strings: {quote-type: double, required: true,
+   skip-quoted-quote: false}``
```
What do you think of this:
```suggestion
#. With ``quoted-strings: {skip-quoted-quote: false}``
```

to make the docs more easily understandable?

rgeraskin replied:
I agree that now it looks a bit overloaded with pre-settings. We could easily drop `required: true` because it wouldn't affect test case logic. But without `quote-type: double` it won't complain about quotes at all.

rgeraskin replied:
I've dropped `required: true` for now.

adrienverge replied:
Right!

adrienverge commented on yamllint/rules/quoted_strings.py:
```diff
@@ -112,6 +115,28 @@
+
+    foo: 'bar"baz'
+
+#. With ``quoted-strings: {quote-type: double, required: true,
+   skip-quoted-quote: true}``
```
```suggestion
#. With ``quoted-strings: {skip-quoted-quote: true}``
```

rgeraskin replied:
I've dropped `required: true` and kept `quote-type: double` because without `quote-type: double` yamllint won't complain about quotes at all.

adrienverge commented on yamllint/rules/quoted_strings.py:
```diff
@@ -125,11 +150,13 @@
         'extra-required': [str],
-        'extra-allowed': [str]}
+        'extra-allowed': [str],
+        'skip-quoted-quote': (True, False)}
```
Better:
```suggestion
        'skip-quoted-quote': bool}
```

rgeraskin replied:
You are right, thanks! Done

adrienverge commented on yamllint/rules/quoted_strings.py:
```diff
@@ -200,19 +229,29 @@ def check(conf, token, prev, next, nextnext, context):
     if (not token.plain) and (token.style == "|" or token.style == ">"):
         return
 
+    # Check value is quoted qoute
```
Can you remove this comment? It's not very useful since the following code creates a var named `is_quoted_quote`.

rgeraskin replied:
done

adrienverge commented on yamllint/rules/quoted_strings.py:
```diff
@@ -235,7 +274,8 @@ def check(conf, token, prev, next, nextnext, context):
 
         # But when used need to match config
-        elif token.style and not _quote_match(quote_type, token.style):
+        elif token.style and not is_quoted_quote and not _quote_match(
+                quote_type, token.style):
```
For better performance (because yamllint can take some time to run on large repositories...), I propose to compute "*does this string contain quoted quote?*" only if needed. This means computing this in last resort (after `_quote_match()`).

Something like:
```suggestion
        elif (token.style and not _quote_match(quote_type, token.style) and
                not (conf['skip-quoted-quote'] and _has_quoted_quotes(token))):
```

rgeraskin replied:
I extracted `is_quoted_quote` variable logic to `_has_quoted_quotes` function and added `conf['allow-quoted-quotes']` check in `if` clauses. Also, I rearranged `if` statements and placed quoted quotes things to the end.

adrienverge replied:
Perfect.

adrienverge reviewed (commented):
Hello @rgeraskin, thanks for contributing!

Can you add tests? Make sure the new tests take example from existing ones and insert nicely, for a full coverage and easier code maintenance.

About the new option name: I propose `allow-quoted-quotes` instead of `skip-quoted-quote`, to use conventional "allow"/"forbid" of yamllint options, and the plural form.

> Closes use-case with quotes inside the string from #424. But still no solution for escape sequences.

What do you mean by "no solution for escape sequences"?

rgeraskin commented:
Hello @adrienverge, thanks for the detailed and fast feedback!

I've added tests to cover all cases for this feature. Also, I agree with you about the feature name, I've renamed it to `allow-quoted-quotes`.

About "no solution for escape sequences". As stated in #424:
> Another case is when quote-type is set to double, but the string contains escape sequences. In those cases, single quote is better, since the content is not interpreted (so we don't need to double backslashes):

So to achieve this functionality yamllint should be able to determine escape sequences in a string and not to complain about wrong quotes even through `quote-type` setting. It's not so easy as the feature from this PR.

adrienverge commented:
Thanks for these modifications. The code in `yamllint/` looks perfect.

However for tests I would really prefer explicit YAML blocks, even if it causes a lot of repetition. It allows easy and reliable visual check of what we expect for correct/wrong YAML code.

Example:
```python
    def test_allow_quoted_quotes(self):
        conf = 'quoted-strings: {required: true, quote-type: '"', allow-quoted-quotes: true}\n'
        self.check('---\n'
                   'foo1: "[barbaz]"\n'
                   'foo2: "[bar'baz]"\n'
                   conf,
                   problem=...)

        conf = 'quoted-strings: {required: false, quote-type: '"', allow-quoted-quotes: true}\n'
        self.check('---\n'
                   'foo1: "[barbaz]"\n'
                   'foo2: "[bar'baz]"\n'
                   conf,
                   problem=...)

        # quote-type: any should accept both:
                   'foo3: '[bar"baz]'\n'
                   'foo3: "[bar'baz]"\n'
	# etc.
```

rgeraskin commented:
@adrienverge I've updated tests for `allow-quoted-quotes` following your advice. Now they looks similar to other tests.

adrienverge reviewed (approved):
Hey Roman, thanks for this cool addition to yamllint, and for your reactivity.

Let's merge this!

adrienverge commented:
I see that [the generated documentation](https://yamllint.readthedocs.io/en/latest/rules.html#module-yamllint.rules.quoted_strings) is wrong, because the first backslash is removed:
![Capture d’écran du 2022-08-07 13-51-52](https://user-images.githubusercontent.com/5244945/183289236-55320312-d25f-42fd-80da-07785947aff7.png)

Do you think you could fix this in a new pull request? (Docs can be generated locally using `python setup.py build_sphinx`.)

rgeraskin commented:
@adrienverge my bad, I've fixed it in #490 

thank you for review!

bendem commented:
Thank you for this, this is great!
